apps/server/src/shared/domain/rules/task.rule.spec.ts

```diff
@@ -29,138 +25,207 @@ describe('TaskRule', () => {
-		const spy = jest.spyOn(lessonRule, 'hasPermission');
-		service.hasPermission(user, entity, { action: Actions.write, requiredPermissions: [permissionA] });
-		expect(spy).toBeCalledWith(user, entity.lesson, { action: Actions.write, requiredPermissions: [] });
-	});
-
-	describe('User [TEACHER]', () => {
-		it('should return "true" if user is creator', () => {
-			entity = taskFactory.build({ creator: user });
-			const res = service.hasPermission(user, entity, { action: Actions.read, requiredPermissions: [] });
-			expect(res).toBe(true);
-		});
-
-		it('should return "true" if user in scope', () => {
-			const course = courseFactory.build({ teachers: [user] });
-			entity = taskFactory.build({ course });
-			const res = service.hasPermission(user, entity, { action: Actions.read, requiredPermissions: [] });
-			expect(res).toBe(true);
-		});
-
-		it('should return "true" if user in scope and lesson hidden', () => {
-			const course = courseFactory.build({ teachers: [user] });
-			const lesson = lessonFactory.build({ course, hidden: true });
-			entity = taskFactory.build({ course, lesson });
-			const res = service.hasPermission(user, entity, { action: Actions.read, requiredPermissions: [] });
-			expect(res).toBe(true);
-		});
-
-		it('should return "false" if user has not permission', () => {
-			entity = taskFactory.build({ creator: user });
-			const res = service.hasPermission(user, entity, { action: Actions.read, requiredPermissions: [permissionC] });
-			expect(res).toBe(false);
-		});
-
-		it('should return "false" if user has not access to entity', () => {
-			entity = taskFactory.build();
-			const res = service.hasPermission(user, entity, { action: Actions.read, requiredPermissions: [permissionC] });
-			expect(res).toBe(false);
-		});
-	});
-
-	describe('User [STUDENT]', () => {
-		let student: User;
-		let role: Role;
-		beforeEach(() => {
-			role = roleFactory.build({ permissions: [permissionA, permissionB], name: RoleName.STUDENT });
-			student = userFactory.build({ roles: [role] });
+	describe('[method] hasPermission', () => {
+		describe('when user with permission A,B and empty task', () => {
+			const setup = () => {
+				const role = roleFactory.build({ permissions: [permissionA, permissionB] });
+				const user = userFactory.build({ roles: [role] });
+				const task = taskFactory.build();
+
+				return { role, user, task };
+			};
+
+			it('should return "false" if user has not access to task', () => {
+				const { user, task } = setup();
+				const res = service.hasPermission(user, task, { action: Actions.read, requiredPermissions: [permissionC] });
+				expect(res).toBe(false);
+			});
 		});
-		it('should return "true" if user is creator', () => {
-			entity = taskFactory.build({ creator: user });
-			const res = service.hasPermission(user, entity, { action: Actions.read, requiredPermissions: [] });
-			expect(res).toBe(true);
+		describe('when user is task creator with permission A,B but not C', () => {
```

Comment:
"permissions A, B, C" are not useful in description, because we don't know what they mean just by reading the description